plugin/pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -17,17 +17,53 @@ limitations under the License.
 package predicates
 
 import (
+	"fmt"
 	"github.com/golang/glog"
 	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/kubernetes/plugin/pkg/scheduler/algorithm"
 	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
 	schedutil "k8s.io/kubernetes/plugin/pkg/scheduler/util"
+	"sync"
 )
 
 type PredicateMetadataFactory struct {
 	podLister algorithm.PodLister
 }
 
+//  Note that predicateMetadata and matchingPodAntiAffinityTerm need to be declared in the same file
+//  due to the way declarations are processed in predicate declaration unit tests.
+type matchingPodAntiAffinityTerm struct {
+	term *v1.PodAffinityTerm
+	node *v1.Node
+}
+
+// NOTE: When new fields are added/removed or logic is changed, please make sure
+// that RemovePod and AddPod functions are updated to work with the new changes.
+type predicateMetadata struct {
+	pod                                *v1.Pod
+	podBestEffort                      bool
+	podRequest                         *schedulercache.Resource
+	podPorts                           map[int]bool
+	matchingAntiAffinityTerms          map[types.UID][]matchingPodAntiAffinityTerm //key is a pod UID with the anti-affinity rule.
+	serviceAffinityInUse               bool
```

Comment:
This bool tells us whether the service affinity predicate is actually enabled. The affinity predicate may be enabled, but the two slices may still be empty because the pod matches no services and no other pods with the same labels exist.